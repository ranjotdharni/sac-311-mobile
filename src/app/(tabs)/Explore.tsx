import { View, StyleSheet, Dimensions } from "react-native";
import MapView, { Details, PROVIDER_GOOGLE, Region } from "react-native-maps";   //By default, this component uses Google Maps as provider
import SearchBar from "../(components)/Profile/SearchBar";
import { global, shadowUniversal, generateEndpointUrl, responseType, dateAtDaysAgo, dateToFormat, generateSymbolUrl } from "../../customs";
import { useEffect, useMemo, useRef, useState } from "react";
import CustomText from "../(components)/CustomText";
import { TouchableOpacity, PanGestureHandler, GestureEvent, PanGestureHandlerEventPayload } from "react-native-gesture-handler";
import Request from "../(components)/Request";
import { FontAwesome } from "@expo/vector-icons";
import CustomMarker from "../(components)/CustomMarker";
import { Loader } from "../(components)/Loader";
import Animated from "react-native-reanimated";
import { FlashList } from "@shopify/flash-list"
import _ from 'lodash'
import { fontGetter } from "../../customs";

//const padKeySuffix: string = '-padRequests'
const markerKeySuffix: string = '-markers'
//const searchKeySuffix: string = '-scrollAddresses'

const padHiddenHeight = Dimensions.get('screen').height
const padVisibleHeight = Dimensions.get('screen').height * 0.47

const requestItemWidth = 0.75 // as a decimal percentage of screen width, used in multiple places synchronously so stored here for maintainability 
//const padAnimationTiming = 1000

function Explore()
{
    const controller = new AbortController()

    const mapRef = useRef<MapView>(null)

    const [addressQuery, setAddressQuery] = useState<string>('')
    const [markers, setMarkers] = useState<Array<responseType>>([])
    const [padDistrict, setPadDistrict] = useState<string>('')
    const [padAddress, setPadAddress] = useState<string>('')
    const [loader, setLoader] = useState<boolean>(false)
    const [addressesLoading, setAddressesLoading] = useState<boolean>(false)
    const [requests, setRequests] = useState<Array<responseType>>([])
    const [data, setData] = useState<Array<responseType>>([])
    const [results, showResults] = useState<boolean>(false)
    const [requestFetch, triggerRequestFetch] = useState<responseType>()
    const [padPan, setPadPan] = useState(padHiddenHeight)
    const [region, setRegion] = useState<Region>({
        latitude: 38.574713,
        longitude: -121.491489,
        latitudeDelta: 0.0922,
        longitudeDelta: 0.0421,
    })

    function hidePad() {
        //setActiveMarker('')
        setPadPan(padHiddenHeight)
    }

    function showPad() {
        setPadPan(padVisibleHeight)
    }

    /* 
        react-native-reanimated was giving compatability issues out the wazoo, fixed the issue for 
        react-native-reanimated and expo complains about compatability, fixed the issue for expo and
        react-native-reanimated complains about compatability; came up with below hacky animation 
        as temporary solution 
    */

    function watchSwipe(evt: GestureEvent<PanGestureHandlerEventPayload>) {
        setPadPan(Math.max(padVisibleHeight, padVisibleHeight + (evt.nativeEvent.translationY / 1.5)))
    }

    function detectSwipeEnd() {
        if (padPan > padVisibleHeight + 100) {
            hidePad()
        }
        else {
            setPadPan(padVisibleHeight)
        }
    }

    //WIP for fetching and parsing the JSON data, uncommenting this makes Expo Go crash every minute or so
    //const [requests, setRequests] = useState([]);
    //fetch('https://services5.arcgis.com/54falWtcpty3V47Z/arcgis/rest/services/SalesForce311_View/FeatureServer/0/query?where=1%3D1&outFields=*&outSR=4326&f=json')
    //.then(response => response.json())
    //.then(json => setRequests(json));

    function initializePad(obj: responseType) { //will wipe current active marker!!!!!!!!!!!!!!
        //setActiveMarker('')
        //setData([])
        //setRequests([])
        showResults(false)
        setPadDistrict(obj.attributes.CouncilDistrictNumber)
        setPadAddress(obj.attributes.Address)
        mapRef.current?.animateToRegion({latitude: obj.geometry.y - 0.005, longitude: obj.geometry.x, latitudeDelta: region.latitudeDelta, longitudeDelta: region.longitudeDelta})
        showPad()
    }   //notice: latitude is geometry.y and longitude is geometry.x

    const fetchAddresses = _.throttle(async () => {
        if (addressQuery === '') {
            setAddressesLoading(false)
            showResults(false)
            return
        }   //break clause

        setAddressesLoading(true)
        showResults(true)

        const query = generateEndpointUrl(`NOT(Address='') AND UPPER(Address) LIKE UPPER('%${addressQuery}%')`, 15, [])

        await fetch(query, { signal: controller.signal }).then((middle) => {
            return middle.json()
        }).then((res) => {
            setAddressesLoading(false)
            setData(res.features)
        }).catch(err => {
            return
            //console.log(err.message)
        })
    }, 750, {leading: false, trailing: true})

    function setQuery(value: string) {
        setAddressQuery(value.trim())
        fetchAddresses()
    }

    function handlePress(obj: responseType) {
        //showResults(false)
        controller.abort()
        setAddressQuery(obj.attributes.Address)
        initializePad(obj)
        setLoader(true)

        setTimeout(() => triggerRequestFetch(obj), 1250)
    }

    function activateMarker(obj: responseType) {
        initializePad(obj)

        //setActiveMarker(obj.attributes.ReferenceNumber + markerKeySuffix)
        setRequests([obj, ...requests])
    }

    const onRegionChange = _.debounce(async (region: Region, details: Details) => {
        // default markers are requests w/ valid addresses and created within in the last 7 days
        const params: [string, string][] = [
            ['orderByFields', 'Address'],
            ['geometryType', 'esriGeometryPoint'],
            ['geometry', `${region.longitude},${region.latitude}`],
            ['distance', '2000'],
            ['returnDistinctValues', 'true']
        ]
        const query = generateEndpointUrl(`NOT(Address='') AND DateCreated > DATE '${dateToFormat('YYYY-MM-DD', dateAtDaysAgo(3))}'`, 75, params)
    
        await fetch(query).then((middle) => {
            return middle.json()
        }).then((res) => {
            setMarkers(res.features)
        }).catch(err => {
            return
            //console.log(err.message)
        })
    }, 1000)

    useEffect(() => {
        async function grabInitialProps() {
            // default markers are requests w/ valid addresses and created within in the last 7 days
            const params: [string, string][] = [
                ['orderByFields', 'Address'],
                ['geometryType', 'esriGeometryPoint'],
                ['geometry', `${region.longitude},${region.latitude}`],
                ['distance', '1000']
            ]
            const query = generateEndpointUrl(`NOT(Address='') AND DateCreated > DATE '${dateToFormat('YYYY-MM-DD', dateAtDaysAgo(7))}'`, 100, params)
        
            await fetch(query).then((middle) => {
                return middle.json()
            }).then((res) => {
                setMarkers(res.features)
            }).catch(err => {
                return
                //console.log(err.message)
            })
        }

        grabInitialProps()
    }, [])

    useEffect(() => {
        async function grabRequests() {
            const query = generateEndpointUrl(`Address='${requestFetch?.attributes.Address}'`, 100, [['orderByFields', 'DateCreated']])

            await fetch(query).then((middle) => {
                return middle.json()
            }).then((res) => {
                setLoader(false)
                setRequests(res.features)
            }).catch(err => {
                return
                //console.log(err.message)
            })
        }

        grabRequests()

    }, [requestFetch])

    const memoizedMarkerRender = useMemo(() => {
        return markers.map((mark, index) => {
            return (
                <CustomMarker 
                    key={mark.attributes.ReferenceNumber + markerKeySuffix} 
                    markerData={mark} 
                    image={generateSymbolUrl(mark.attributes.CategoryLevel1)}
                    iconScale={25}
                    fadeInDelay={index * 50}
                    passUp={activateMarker}
                />
            )
        })
    }, [markers])

    /*const memoizedRequestsRender = useCallback(({item} : {item: any}) => 
        <Request
            data={item}
            width={Dimensions.get('screen').width * 0.75}
            height={Dimensions.get('screen').height * 0.1}
            compact={true}
    />, [])*/

    const memoizedRequestData = useMemo(() => requests.map((req, idx) => {
        return (
            <Request 
                key={idx}
                data={req}
                width={Dimensions.get('screen').width * requestItemWidth}
                height={Dimensions.get('screen').height * 0.1}
                compact={true}
            />
        )
    }), [requests])
    //req.attributes.ReferenceNumber + padKeySuffix

    const memoizedAddressData = useMemo(() => data.map((addr, idx) => {
        return (
            <TouchableOpacity key={idx} onPress={ () => handlePress(addr) } style={[styles.resultShadow, shadowUniversal.default]}>
                <CustomText text={addr.attributes.Address} style={styles.result} nol={0} font={fontGetter()} />
            </TouchableOpacity>
        )
    }), [data])

    return (
        <View style={{flex: 1}}>
                <Animated.View style={[styles.requestWindow, { transform: [{ translateY: padPan }] }]}>
                    <PanGestureHandler onEnded={detectSwipeEnd} onGestureEvent={watchSwipe}>
                        <View style={styles.padTopBar}>
                            <View style={styles.closeBar}></View>
                            
                            <View style={styles.padLeftPartition}>
                                <CustomText nol={3} text={padAddress} font={fontGetter()} style={styles.padAddress} />
                                <View style={{flexDirection: 'row', marginBottom: '1%'}}>
                                    <FontAwesome name='trash' size={30} color={global.baseGrey200} />
                                    <CustomText text='Thursday' nol={0} font={fontGetter()} style={{marginTop: '4%', marginLeft: '2.5%', color: global.baseGrey200}} />
                                </View>
                            </View>

                            <View style={styles.padRightPartition}>
                                <CustomText text={padDistrict} nol={0} font={fontGetter()} style={styles.padDistrict} />
                                <TouchableOpacity style={styles.newRequestButton}>
                                    <CustomText text='New Request' nol={0} font={fontGetter()} style={{fontSize: 15, padding: 15, color: 'white', textAlign: 'center'}} />
                                </TouchableOpacity>
                            </View>
                        </View>
                    </PanGestureHandler>
                    {
                        loader ? 
                        <Loader /> :
                        <FlashList
                            contentContainerStyle={{paddingLeft: Dimensions.get('screen').width * ((1 - requestItemWidth) / 4), paddingBottom: Dimensions.get('screen').height * 0.05}}
                            data={memoizedRequestData}
                            renderItem={({item}) => item}
                            estimatedItemSize={100}
                        />
                    }
                </Animated.View>
                <Animated.View style={{width: '100%', height: '100%'}}>
                <MapView onRegionChange={onRegionChange} ref={mapRef} provider={PROVIDER_GOOGLE} region={region} style={{width: '100%', height: '100%', position:'absolute'}}>
                    {
                        memoizedMarkerRender
                    }
                </MapView>
                </Animated.View>
                <SearchBar value={addressQuery} style={styles.searchBar} onSubmit={() => { showResults(false) }} passUp={setQuery} onClear={() => { setAddressQuery('') }} placeholder={'Search Address'} />
                {
                    (
                        results ?
                        <View style={[styles.searchResults, shadowUniversal.default]}>
                            {
                                (
                                    addressesLoading ? 
                                    <Loader /> :
                                    <FlashList 
                                        estimatedItemSize={15}
                                        data={memoizedAddressData}
                                        renderItem={({item}) => item}
                                    />
                                )
                            }
                        </View> :
                        <></>
                    )
                }
        </View>
    )
}

const styles = StyleSheet.create({
    wrapper: {
        flex: 1,
    },

    map: {
        width: '100%',
        height: '100%',
    },

    searchBar: {
        width: '95%',
        position: 'absolute',
        zIndex: 2,
        top: '5%',
        left: '2.5%',
        borderWidth: 1,
        borderColor: global.baseBlue100,
    },

    searchResults: {
        backgroundColor: global.baseBackground100,
        width: '90%',
        height: '40%',
        position: 'absolute',
        top: '13.5%',
        left: '5%',
        borderRadius: 15,
        zIndex: 5,
    },

    searchResultsPaddingBottom: {
        width: '100%',
        height: 25,
    },

    resultShadow: {
        backgroundColor: global.baseBackground100,
        borderRadius: 10,
        alignSelf: 'center',
        marginTop: '2.5%',
        padding: '5%',
        width: '90%',
    },

    result: {
        textAlign: 'center',
        overflow: 'hidden',
    },

    requestWindow: {
        backgroundColor: global.baseBackground100,
        borderRadius: 15,
        alignSelf: 'center',
        width: '88%',
        height: '40%',
        overflow: 'scroll',
        zIndex: 3,
        position: 'absolute',
        borderColor: 'transparent',
        shadowColor: "#000000",
        shadowOffset: {
	        width: 0,
	        height: 5,
        },
        shadowOpacity: 0.30,
        shadowRadius: 6,
        elevation: 9,
    },

    padTopBar: {
        backgroundColor: global.baseBackground100,
        width: '90%',
        height: '35%',
        left: '5%',
        borderBottomColor: '#1b1b1b2f',
        borderBottomWidth: 2,
        display: 'flex',
        flexDirection: 'row',
    },

    closeBar: {
        position: 'absolute',
        width: '25%',
        height: 5,
        top: 5,
        left: '37.5%',
        borderRadius: 15,
        backgroundColor: global.baseGrey200,
    },

    requestBubble: {
        backgroundColor: '#c79d3c',
        color: '#ffffff',
        opacity: 1,
        borderRadius: 15,
        alignSelf: 'center',
        width: '90%',
        height: 'auto',
        padding:5,
        margin:3,
        shadowColor: "#000000",
        shadowOffset: {
	        width: 0,
	        height: 5,
        },
        shadowOpacity: 0.20,
        shadowRadius: 6,
        elevation: 1,
    },

    requestText: {
        color: '#ffffff',
        left: "5%",
        fontSize: 14
    },

    padLeftPartition: {
        width: '60%',
        height: '100%',
        flexDirection: 'column',
        justifyContent: 'space-between',
    },

    padRightPartition: {
        width: '40%',
        height: '100%',
        flexDirection: 'column',
        justifyContent: 'space-between',
    },

    padAddress: {
        marginTop: '10%',
        color: global.baseGold100,
        fontSize: 18.5,
    },

    padDistrict: {
        marginTop: '16%',
        marginRight: '10%',
        color: global.baseGrey100,
        fontSize: 15,
        alignSelf: 'flex-end'
    },

    newRequestButton: {
        backgroundColor: global.baseBlue100,
        borderRadius: 30,
        marginBottom: '5%',
    }
})

export default Explore