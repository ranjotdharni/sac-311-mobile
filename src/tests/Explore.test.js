import renderer from "react-test-renderer";
import Explore from "../app/(tabs)/Explore";

describe("Explore", () => {
  it("renders correctly", () => {
    const page = renderer.create(<Explore />).toJSON();
    expect(page).toMatchSnapshot();
  });
});