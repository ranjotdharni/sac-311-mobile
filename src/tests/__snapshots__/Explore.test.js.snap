// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explore renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#f5faff",
        "borderRadius": 5,
        "display": "flex",
        "elevation": 5,
        "flexDirection": "row",
        "height": "4%",
        "justifyContent": "space-between",
        "left": "75%",
        "position": "absolute",
        "shadowColor": "black",
        "shadowOffset": {
          "height": 4,
          "width": -2,
        },
        "shadowOpacity": 0.33,
        "shadowRadius": 10,
        "transform": [
          {
            "translateY": 1133.8999999999999,
          },
        ],
        "width": "18%",
        "zIndex": 2,
      }
    }
  >
    <Text />
    <View
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "left": "50%",
          "position": "absolute",
          "width": "50%",
        }
      }
    />
  </View>
  <View
    collapsable={false}
    style={
      {
        "alignSelf": "center",
        "backgroundColor": "#f5faff",
        "borderColor": "transparent",
        "borderRadius": 15,
        "elevation": 9,
        "height": "40%",
        "overflow": "scroll",
        "position": "absolute",
        "shadowColor": "#000000",
        "shadowOffset": {
          "height": 5,
          "width": 0,
        },
        "shadowOpacity": 0.3,
        "shadowRadius": 6,
        "transform": [
          {
            "translateY": 1334,
          },
        ],
        "width": "88%",
        "zIndex": 3,
      }
    }
  >
    <View
      collapsable={false}
      handlerTag={1}
      handlerType="PanGestureHandler"
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      style={
        {
          "backgroundColor": "#f5faff",
          "borderBottomColor": "#1b1b1b2f",
          "borderBottomWidth": 2,
          "display": "flex",
          "flexDirection": "row",
          "height": "35%",
          "left": "5%",
          "width": "90%",
        }
      }
    >
      <View
        style={
          {
            "flexDirection": "column",
            "height": "100%",
            "justifyContent": "space-between",
            "width": "60%",
          }
        }
      >
        <View
          style={
            {
              "flexDirection": "row",
              "marginBottom": "1%",
            }
          }
        >
          <Text />
        </View>
      </View>
      <View
        style={
          {
            "flexDirection": "column",
            "height": "100%",
            "justifyContent": "space-between",
            "width": "40%",
          }
        }
      >
        <RNGestureHandlerButton
          collapsable={false}
          delayLongPress={600}
          enabled={true}
          exclusive={true}
          handlerTag={2}
          handlerType="NativeViewGestureHandler"
          onGestureEvent={[Function]}
          onGestureHandlerEvent={[Function]}
          onGestureHandlerStateChange={[Function]}
          onHandlerStateChange={[Function]}
          rippleColor={0}
          touchSoundDisabled={false}
        >
          <View
            accessible={true}
            collapsable={false}
            style={
              {
                "backgroundColor": "#203d61",
                "borderRadius": 30,
                "marginBottom": "5%",
                "opacity": 1,
              }
            }
          />
        </RNGestureHandlerButton>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#203d61",
            "borderRadius": 10,
            "display": "flex",
            "flexDirection": "row",
            "height": 20,
            "justifyContent": "center",
            "left": "97.5%",
            "position": "absolute",
            "top": "5%",
            "width": 20,
          }
        }
      >
        <Text />
      </View>
    </View>
    <View
      style={
        {
          "flex": 1,
          "overflow": "hidden",
        }
      }
    >
      <RCTScrollView
        applyWindowCorrection={[Function]}
        canChangeSize={true}
        contentContainerStyle={
          {
            "backgroundColor": undefined,
            "minHeight": 1,
            "minWidth": 1,
            "paddingLeft": 46.875,
            "paddingRight": 0,
          }
        }
        contentHeight={0}
        contentWidth={0}
        data={[]}
        dataProvider={
          DataProvider {
            "_data": [],
            "_firstIndexToProcess": 0,
            "_hasStableIds": false,
            "_requiresDataChangeHandling": false,
            "_size": 0,
            "getStableId": [Function],
            "rowHasChanged": [Function],
          }
        }
        disableRecycling={false}
        estimatedItemSize={100}
        extendedState={{}}
        externalScrollView={[Function]}
        finalRenderAheadOffset={250}
        forceNonDeterministicRendering={true}
        horizontal={false}
        initialOffset={0}
        initialRenderIndex={0}
        isHorizontal={false}
        layoutProvider={
          GridLayoutProviderWithProps {
            "_acceptableRelayoutDelta": 1,
            "_getHeightOrWidth": [Function],
            "_getLayoutTypeForIndex": [Function],
            "_getSpan": [Function],
            "_hasExpired": false,
            "_maxSpan": 1,
            "_setLayoutForType": [Function],
            "_tempDim": {
              "height": 0,
              "width": 0,
            },
            "averageWindow": AverageWindow {
              "currentAverage": 100,
              "currentCount": 1,
              "inputValues": [
                100,
              ],
              "nextIndex": 1,
            },
            "defaultEstimatedItemSize": 100,
            "layoutObject": {
              "size": undefined,
              "span": undefined,
            },
            "props": {
              "contentContainerStyle": {
                "paddingBottom": 66.7,
                "paddingLeft": 46.875,
              },
              "data": [],
              "estimatedItemSize": 100,
              "numColumns": 1,
              "renderItem": [Function],
            },
            "renderWindowInsets": {
              "height": 0,
              "width": -46.875,
            },
            "shouldRefreshWithAnchoring": true,
          }
        }
        maxRenderAhead={750}
        numColumns={1}
        onEndReached={[Function]}
        onEndReachedThreshold={0}
        onEndReachedThresholdRelative={0}
        onItemLayout={[Function]}
        onLayout={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onSizeChanged={[Function]}
        onVisibleIndicesChanged={[Function]}
        removeClippedSubviews={false}
        renderAheadOffset={0}
        renderAheadStep={250}
        renderContentContainer={[Function]}
        renderItem={[Function]}
        renderItemContainer={[Function]}
        rowRenderer={[Function]}
        scrollEventThrottle={16}
        scrollThrottle={16}
        scrollViewProps={
          {
            "contentContainerStyle": {
              "backgroundColor": undefined,
              "minHeight": 1,
              "minWidth": 1,
              "paddingLeft": 46.875,
              "paddingRight": 0,
            },
            "onLayout": [Function],
            "onScrollBeginDrag": [Function],
            "refreshControl": undefined,
            "style": {
              "minHeight": 1,
              "minWidth": 1,
            },
          }
        }
        style={
          {
            "minHeight": 1,
            "minWidth": 1,
          }
        }
        suppressBoundedSizeException={true}
        windowCorrectionConfig={
          {
            "applyToInitialOffset": true,
            "applyToItemScroll": true,
            "value": {
              "endCorrection": 0,
              "startCorrection": 0,
              "windowShift": -0,
            },
          }
        }
      >
        <View>
          <View
            style={
              {
                "flexDirection": "column",
              }
            }
          >
            <View
              style={
                {
                  "paddingLeft": 46.875,
                  "paddingTop": 0,
                }
              }
            />
            <View
              style={
                [
                  undefined,
                  undefined,
                ]
              }
            />
            <AutoLayoutView
              enableInstrumentation={false}
              horizontal={false}
              onBlankAreaEvent={[Function]}
              onLayout={[Function]}
              renderAheadOffset={0}
              scrollOffset={0}
              style={
                {
                  "height": 0,
                  "width": 0,
                }
              }
              windowSize={0}
            />
            <CellContainer
              index={-1}
              style={
                [
                  undefined,
                  undefined,
                ]
              }
            />
            <View
              style={
                {
                  "paddingBottom": 66.7,
                  "paddingRight": 0,
                }
              }
            />
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
  <View
    collapsable={false}
    style={
      {
        "height": "100%",
        "width": "100%",
      }
    }
  />
  <View
    style={
      [
        {
          "backgroundColor": "red",
          "borderRadius": 25,
          "display": "flex",
          "flexDirection": "column",
          "height": 50,
          "justifyContent": "center",
          "marginTop": 10,
          "overflow": "hidden",
          "shadowColor": "#000",
          "shadowOffset": {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 3.84,
          "width": 750,
        },
        {
          "borderColor": "#203d61",
          "borderWidth": 1,
          "height": 40,
          "left": "2.5%",
          "position": "absolute",
          "top": "5%",
          "width": "95%",
          "zIndex": 2,
        },
      ]
    }
  >
    <View
      style={
        {
          "backgroundColor": "white",
          "borderRadius": 25,
          "elevation": 10,
          "height": 50,
          "width": 750,
        }
      }
    >
      <Image
        source={1}
        style={
          {
            "height": 20,
            "left": 10,
            "position": "absolute",
            "top": 15,
            "width": 20,
          }
        }
      />
      <Text
        style={
          {
            "color": "#D3D3D3",
            "fontFamily": "jbm",
            "left": 40,
            "position": "absolute",
            "top": 15,
          }
        }
      >
        Search Address
      </Text>
      <TextInput
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        placeholderTextColor="#D3D3D3"
        style={
          {
            "backgroundColor": "rgba(0, 0, 0, 0)",
            "borderRadius": 25,
            "fontFamily": "jbm",
            "height": "100%",
            "paddingLeft": 40,
            "paddingRight": 10,
          }
        }
        value=""
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#f5faff",
            "display": "flex",
            "flexDirection": "column",
            "height": "50%",
            "justifyContent": "center",
            "left": "82.5%",
            "opacity": 1,
            "position": "absolute",
            "top": "25%",
            "width": "10%",
            "zIndex": 2,
          }
        }
      >
        <Text />
      </View>
    </View>
  </View>
</View>
`;
